json/src/main/scala/spray/json/JsonParser.scala

```diff
@@ -0,0 +1,347 @@
+      case '"' =>
+        `string`(); jsValue = if (sb.length == 0) JsString.empty else JsString(sb.toString)
+      case _ => fail("JSON Value")
+    }
+  }
+
+  private def `false`() = advance() && ch('a') && ch('l') && ch('s') && ws('e')
+  private def `null`() = advance() && ch('u') && ch('l') && ws('l')
+  private def `true`() = advance() && ch('r') && ch('u') && ws('e')
+
+  // http://tools.ietf.org/html/rfc4627#section-2.2
+  private def `object`(): Unit = {
+    ws()
+    jsValue = if (cursorChar != '}') {
+      @tailrec def members(map: Map[String, JsValue]): Map[String, JsValue] = {
+        `string`()
+        require(':')
+        ws()
+        val key = sb.toString
+        `value`()
+        val nextMap = map.updated(key, jsValue)
+        if (ws(',')) members(nextMap) else nextMap
+      }
+      var map = Map.empty[String, JsValue]
+      map = members(map)
+      require('}')
+      JsObject(map)
+    } else {
+      advance()
+      JsObject.empty
+    }
+    ws()
+  }
+
+  // http://tools.ietf.org/html/rfc4627#section-2.3
+  private def `array`(): Unit = {
+    ws()
+    jsValue = if (cursorChar != ']') {
+      val list = Vector.newBuilder[JsValue]
+      @tailrec def values(): Unit = {
+        `value`()
+        list += jsValue
+        if (ws(',')) values()
+      }
+      values()
+      require(']')
+      JsArray(list.result())
+    } else {
+      advance()
+      JsArray.empty
+    }
+    ws()
+  }
+
+  // http://tools.ietf.org/html/rfc4627#section-2.4
+  private def `number`() = {
+    val start = input.cursor
+    val startChar = cursorChar
+    ch('-')
+    `int`()
+    `frac`()
+    `exp`()
+    jsValue =
+      if (startChar == '0' && input.cursor - start == 1) JsNumber.zero
+      else JsNumber(input.sliceCharArray(start, input.cursor))
+    ws()
+  }
+
+  private def `int`(): Unit = if (!ch('0')) oneOrMoreDigits()
+  private def `frac`(): Unit = if (ch('.')) oneOrMoreDigits()
+  private def `exp`(): Unit = if (ch('e') || ch('E')) { ch('-') || ch('+'); oneOrMoreDigits() }
+
+  private def oneOrMoreDigits(): Unit = if (DIGIT()) zeroOrMoreDigits() else fail("DIGIT")
+  @tailrec private def zeroOrMoreDigits(): Unit = if (DIGIT()) zeroOrMoreDigits()
+
+  private def DIGIT(): Boolean = cursorChar >= '0' && cursorChar <= '9' && advance()
+
+  // http://tools.ietf.org/html/rfc4627#section-2.5
+  private def `string`(): Unit = {
+    if (cursorChar == '"') cursorChar = input.nextUtf8Char() else fail("'\"'")
+    sb.setLength(0)
+    while (`char`()) cursorChar = input.nextUtf8Char()
+    require('"')
+    ws()
+  }
+
+  private def `char`() =
+    // simple bloom-filter that quick-matches the most frequent case of characters that are ok to append
+    // (it doesn't match control chars, EOI, '"', '?', '\', 'b' and certain higher, non-ASCII chars)
+    if (((1L << cursorChar.toInt) & ((31L - cursorChar.toLong) >> 31) & 0x7ffffffbefffffffL) != 0L) appendSB(cursorChar)
```

Comment:
It's a micro-optimization reducing the most common code path to a single branch instruction. Its effect is probably barely noticeable, but you have to admit that it looks cool... :)